### PR TITLE
ENH: between_time, at_time accept axis parameter

### PR DESCRIPTION
**Pull request by yrhooke**
xref #2141 

- [x] closes #8839
- [x] tests added / passed

Axis parameter added to between_time and at_time, defaults to index for DataFrame

### CODE REVIEW

File: pandas/core/generic.py [hunk 1]
```diff
@@ -6772,6 +6772,7 @@ def at_time(self, time, asof=False):
         Parameters
         ----------
         time : datetime.time or string
+        axis : int or string axis name, optional
```
Comment (jreback) [inline pandas/core/generic.py]: can you add a versionadded 0.24.0 here (and other new kwargs)

File: pandas/core/generic.py [hunk 2]
```diff
@@ -6826,6 +6832,7 @@ def between_time(self, start_time, end_time, include_start=True,
         end_time : datetime.time or string
         include_start : boolean, default True
         include_end : boolean, default True
+        axis : int or string axis name, optional
```
Comment (jreback) [inline pandas/core/generic.py]: same

File: pandas/tests/frame/test_timeseries.py [hunk 3]
```diff
@@ -706,6 +722,37 @@ def test_between_time_raises(self):
         with pytest.raises(TypeError):  # index is not a DatetimeIndex
             df.between_time(start_time='00:00', end_time='12:00')
 
+    def test_between_time_axis(self):
+        rng = date_range('1/1/2000', periods=100, freq='10min')
+        blank = np.arange(0, len(rng))
+        stime, etime = ('08:00:00', '09:00:00')
+        dimn = (len(rng), len(rng))
+        exp_len = 7
+        for time_index, time_col in product([True, False], [True, False]):
```
Comment (jreback) [inline pandas/tests/frame/test_timeseries.py]: can you parametrize this test instead of using a loop

File: pandas/tests/frame/test_timeseries.py [hunk 4]
```diff
@@ -640,6 +640,22 @@ def test_at_time_raises(self):
         with pytest.raises(TypeError):  # index is not a DatetimeIndex
             df.at_time('00:00')
 
+    def test_at_time_axis(self):
```
Comment (jreback) [inline pandas/tests/frame/test_timeseries.py]: you can parametrize if you want

File: pandas/tests/series/test_timeseries.py [hunk 5]
```diff
@@ -817,6 +817,16 @@ def test_between_time_formats(self):
         for time_string in strings:
             assert len(ts.between_time(*time_string)) == expected_length
 
+    def test_between_time_axis(self):
+        rng = date_range('1/1/2000', periods=100, freq='10min')
```
Comment (jreback) [inline pandas/tests/series/test_timeseries.py]: can you add the issue number as a comment (to all new tests)

**Review by jreback — CHANGES_REQUESTED** [review]
can you add a whatsnew in other enhancements for 0.24.0

**Comment by jreback** [comment]
a nice followup would be #2141

**Comment by yrhooke** [comment]
Alright, I'm on it.

File: doc/source/whatsnew/v0.24.0.txt [hunk 6]
```diff
@@ -114,6 +114,7 @@ Other Enhancements
 - :meth:`Series.nlargest`, :meth:`Series.nsmallest`, :meth:`DataFrame.nlargest`, and :meth:`DataFrame.nsmallest` now accept the value ``"all"`` for the ``keep`` argument. This keeps all ties for the nth largest/smallest value (:issue:`16818`)
 - :class:`IntervalIndex` has gained the :meth:`~IntervalIndex.set_closed` method to change the existing ``closed`` value (:issue:`21670`)
 - :func:`~DataFrame.to_csv` and :func:`~DataFrame.to_json` now support ``compression='infer'`` to infer compression based on filename (:issue:`15008`)
+- :func:`between_time` and :func:`at_time` now support an axis parameter (:issue: `8839`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.24.0.txt]: can you use double backticks around axis

File: pandas/core/generic.py [hunk 7]
```diff
@@ -6950,14 +6954,19 @@ def at_time(self, time, asof=False):
...
             values at particular time of the day
         """
+        if axis is None:
+            axis = self._stat_axis_number
+        axis = self._get_axis_number(axis)
+
         try:
-            indexer = self.index.indexer_at_time(time, asof=asof)
-            return self._take(indexer)
+            index = self._get_axis(axis)
```
Comment (jreback) [inline pandas/core/generic.py]: can you move the _get_axis out of the try/except (and the _take)

File: pandas/core/generic.py [hunk 8]
```diff
@@ -7012,11 +7024,16 @@ def between_time(self, start_time, end_time, include_start=True,
         DatetimeIndex.indexer_between_time : Get just the index locations for
             values between particular times of the day
         """
+        if axis is None:
+            axis = self._stat_axis_number
+        axis = self._get_axis_number(axis)
+
         try:
-            indexer = self.index.indexer_between_time(
+            index = self._get_axis(axis)
```
Comment (jreback) [inline pandas/core/generic.py]: same comment as above

File: pandas/tests/frame/test_timeseries.py [hunk 9]
```diff
@@ -640,6 +640,27 @@ def test_at_time_raises(self):
         with pytest.raises(TypeError):  # index is not a DatetimeIndex
             df.at_time('00:00')
 
+    @pytest.mark.parametrize('time_axis', [
+        (False, False), (True, False), (False, True), (True, True)])
```
Comment (jreback) [inline pandas/tests/frame/test_timeseries.py]: hmm this is an odd way to parameterize, can you just use axis as the argument?

File: pandas/tests/frame/test_timeseries.py [hunk 10]
```diff
@@ -706,6 +727,40 @@ def test_between_time_raises(self):
         with pytest.raises(TypeError):  # index is not a DatetimeIndex
             df.between_time(start_time='00:00', end_time='12:00')
 
+    @pytest.mark.parametrize('time_axis', [
```
Comment (jreback) [inline pandas/tests/frame/test_timeseries.py]: same this is very hard to read

Comment (jreback) [reply pandas/tests/frame/test_timeseries.py]: you can make the error case a separate test, which makes both tests much simpler

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by yrhooke** [comment]
I hope it's alright now. 

`test_between_time_axis` and `test_between_time_axis_raises` are trying to account for the scenarios where neither axis is a DatetimeIndex or that both are. If those are extraneous I can simplify them. 

Also, should these tests be combined into the main `test_between_time`, `test_at_time`? I'm not sure what's preferred here.

File: pandas/tests/frame/test_timeseries.py [hunk 11]
```diff
@@ -640,6 +640,23 @@ def test_at_time_raises(self):
         with pytest.raises(TypeError):  # index is not a DatetimeIndex
             df.at_time('00:00')
 
+    @pytest.mark.parametrize('axis', ['index', 'columns'])
```
Comment (jreback) [inline pandas/tests/frame/test_timeseries.py]: test for 0, 1 as well

File: pandas/tests/frame/test_timeseries.py [hunk 12]
```diff
@@ -706,6 +723,44 @@ def test_between_time_raises(self):
         with pytest.raises(TypeError):  # index is not a DatetimeIndex
             df.between_time(start_time='00:00', end_time='12:00')
 
+    @pytest.mark.parametrize('axis', [
```
Comment (jreback) [inline pandas/tests/frame/test_timeseries.py]: also test for 0, 1

**Review by jreback — CHANGES_REQUESTED** [review]
small comments. pls rebase and ping on green.

**Comment by jreback** [comment]
can you rebase and add those tests, otherwise lgtm.

File: pandas/tests/frame/test_timeseries.py (hunk 12 above)
Comment (jreback) [reply pandas/tests/frame/test_timeseries.py]: the new axis fixture will handle this, can you use that

Comment (yrhooke) [reply pandas/tests/frame/test_timeseries.py]: which new axis fixture?

Comment (jreback) [reply pandas/tests/frame/test_timeseries.py]: rebase in master and axis is available

**Comment by jreback** [comment]
can you rebase and update

**Comment by yrhooke** [comment]
I rebased and updated the tests, they're kind of awkwardly built, I couldn't figure out what the axis fixture was.

File: doc/source/whatsnew/v0.24.0.txt [hunk 13]
```diff
@@ -182,6 +182,9 @@ Other Enhancements
 - :func:`to_timedelta` now supports iso-formated timedelta strings (:issue:`21877`)
 - :class:`Series` and :class:`DataFrame` now support :class:`Iterable` in constructor (:issue:`2193`)
 - :class:`DatetimeIndex` gained :attr:`DatetimeIndex.timetz` attribute. Returns local time with timezone information. (:issue:`21358`)
+- :func:`~DataFrame.to_csv` and :func:`~DataFrame.to_json` now support ``compression='infer'`` to infer compression based on filename (:issue:`15008`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.24.0.txt]: looks like you picked up something extra here

File: pandas/tests/frame/test_timeseries.py [hunk 14]
```diff
@@ -640,6 +640,23 @@ def test_at_time_raises(self):
         with pytest.raises(TypeError):  # index is not a DatetimeIndex
             df.at_time('00:00')
 
+    @pytest.mark.parametrize('axis', ['index', 'columns', 0, 1])
```
Comment (jreback) [inline pandas/tests/frame/test_timeseries.py]: just remove the paramerize, the axis is already defined

File: pandas/tests/frame/test_timeseries.py (hunk 12 above)
Comment (jreback) [reply pandas/tests/frame/test_timeseries.py]: why is this not just using the axis fixture?

File: pandas/tests/frame/test_timeseries.py [hunk 15]
```diff
@@ -706,6 +723,46 @@ def test_between_time_raises(self):
...
+            assert len(ts.between_time(stime, etime)) == exp_len
+            assert len(ts.between_time(stime, etime, axis=0)) == exp_len
+
+        if 'columns' in axis or 1 in axis:
+            ts.columns = rng
+            selected = ts.between_time(stime, etime, axis=1).columns
+            assert len(selected) == exp_len
+
+    @pytest.mark.parametrize('axis', [
+        (), ('index',), ('columns',), ('index', 'columns'),
```
Comment (jreback) [inline pandas/tests/frame/test_timeseries.py]: same, this is very werid

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/tests/frame/test_timeseries.py (hunk 12 above)
Comment (yrhooke) [reply pandas/tests/frame/test_timeseries.py]: I couldn't figure out what the axis fixture is. Is there documentation about this?

Comment (WillAyd) [reply pandas/tests/frame/test_timeseries.py]: Here's the location of it. It just supplies the various arguments that are typically valid for an `axis` parameter

https://github.com/pandas-dev/pandas/blob/c04b7739b7caa04496c876b19c37e0decf88af78/pandas/conftest.py#L65

**Comment by jreback** [comment]
think this got lost, can you rebase

**Comment by jreback** [comment]
can you merge master

**Comment by jreback** [comment]
ok i rebased this.

**Review by jreback — APPROVED** [review]

**Pull request merged**